Fixing SizeF, CookieContainer, CookieCollection serialization between Core <--> Desktop and adding full binary mode tests

ViktorHofer opened the pull request:
Relates to https://github.com/dotnet/coreclr/pull/12365
Port for 2.0.0: https://github.com/dotnet/corefx/pull/21224

## Fixing
- SizeF public key token was wrong
- CookieContainer core --> desktop serialization in simple and full mode
- CookieCollection serialization core <--> desktop in simple and full mode (A regression introduced by my latest change on Friday)

## Adding
- Tests for full mode binaryformatter
- Integrity test for blobs

## Requires
- CompareInfo culture field serialization fix in coreclr: https://github.com/dotnet/coreclr/pull/12365

What I also realized, we aren't updating the blobs in CI which means that we don't detect if a change is harming our serialization support...

cc @danmosemsft @krwq

danmosemsft commented:
> What I also realized, we aren't updating the blobs in CI which means that we don't detect if a change is harming our serialization support...

Can you explain this? I would expect not updating the blobs in CI to be a good thing, in that code changes still have to work with those known good blobs?

danmosemsft commented on src/System.Runtime.Serialization.Formatters/tests/BinaryFormatterHelper.cs:
```diff
@@ -199,35 +202,37 @@ public static void UpdateCoreTypeBlobs(string testDataFilePath, string[] blobs)
         }
 
-        public static byte[] SerializeObjectToRaw(object obj)
+        public static byte[] SerializeObjectToRaw(object obj, FormatterAssemblyStyle assemblyStyle = FormatterAssemblyStyle.Simple)
```
Why default values? Is it worth being explicit in all the callers, so it's clear?

ViktorHofer replied:
We can do that. Don't have a strong opinion on that.

krwq replied:
I think no default is better here since we are testing this we always want to check all values

danmosemsft commented on src/System.Net.Primitives/src/System/Net/CookieCollection.cs:
```diff
@@ -24,7 +24,8 @@ internal enum Stamp
-        private DateTime m_timeStamp = DateTime.MinValue; // Do not rename (binary serialization)
+        private int m_version; // Do not rename (binary serialization)
+        private DateTime m_TimeStamp = DateTime.MinValue; // Do not rename (binary serialization)
         private bool m_has_other_versions; // Do not rename (binary serialization)
```
Desktop has
```
[OptionalField]
bool m_IsReadOnly;
```
This means it's OK to deserialize to Desktop a stream without this field. But does it mean that it's going to work to deserialize to Core a stream that DOES have this field? The documentation doesn't say it is. Perhaps you need this optional field also.

ViktorHofer replied:
OptionalFields are not required and would fail in the current tests now.

krwq replied:
@ViktorHofer what about netfx->core? If netfx doesn't create this value then core deserialization will fail

ViktorHofer replied:
About which value are we talking now? `m_IsReadOnly` or `m_version`? 

- m_IsReadOnly doesn't exist in Core as we don't need it and it is marked in at least netfx >= 461 as optional.
- m_version should always be in the netfx blob. Why shouldn't it?

danmosemsft replied:
The question is what we do when a payload arrives at Core that DOES contain m_IsReadOnly (which we don't have a field for). Does the deserializer just ignore the unexpected extra value? If we are currently testing with a 4.6.1 blob that does contain this value, then it's clearly working and we're fine.

ViktorHofer commented:
If someone is still up after a new coreclr maestro comes in, then please retrigger the tests here and merge if green... I just merged the dependent coreclr change so it might take a while as I can't new coreclr builds

danmosemsft commented:
@dotnet-bot test this please (Retesting with new coreclr)

@ViktorHofer note the "portable..." legs are dead, ignore those now.

danmosemsft commented:
Bunch of tests failing against Desktop.
https://ci.dot.net/job/dotnet_corefx/job/master/job/netfx_windows_nt_debug_prtest/1831/
`System.Runtime.Serialization.SerializationException : Member 'culture' in class 'System.Globalization.CompareInfo' is not present in the serialized stream and is not marked with System.Runtime.Serialization.OptionalFieldAttribute.`

This is deserializing onto desktop presumably? On desktop I see it, but also another field that is optional
```
                [OptionalField(VersionAdded = 1)]
                private int win32LCID;             // mapped sort culture id of this instance
                private int culture;               // the culture ID used to create this instance
```

https://referencesource.microsoft.com/#mscorlib/system/globalization/compareinfo.cs,258

Not sure what's going on. In theory coreclr is not involved in the desktop tests.

danmosemsft commented:
@ViktorHofer ?

ViktorHofer commented:
The core blobs need to be updated to now contain the culture field. I will do that

ViktorHofer commented:
I verified the fix locally in the release branches coreclr-->corefx, updated the blobs here and in the release PR.

stephentoub commented on src/System.Net.Primitives/src/System/Net/CookieCollection.cs:
```diff
@@ -64,6 +65,7 @@ public void Add(Cookie cookie)
             {
                 throw new ArgumentNullException(nameof(cookie));
             }
+            ++m_version;
```
Is this actually needed?

ViktorHofer replied:
Not in Core but in netfx therefore we should serialize the correct value.

stephentoub replied:
Why does the value matter?  Isn't it only used to determine whether an item has been added during enumeration of the collection?

ViktorHofer replied:
> Isn't it only used to determine whether an item has been added during enumeration of the collection?

Correct but we need the field for correct Desktop <--> Core serialization in FormatterAssemblyMode.Full. I don't really care if we remove it or not, I just don't think that removing the increment operation gives us noteworthy performance benefits. I would just stick with the netfx behavior...

ViktorHofer replied:
But don't get me wrong, I totally understand that we don't really need the correct value in the field for compatibility with netfx. I'm just not sure if that's also true for earlier versions or other frameworks like mono.

stephentoub replied:
> But don't get me wrong, I totally understand that we don't really need the correct value in the field for compatibility with netfx. I'm just not sure if that's also true for earlier versions or other frameworks like mono.

It's very strange to me that we're caring about the value we set it to and then never reading it.  If we were using it in the enumerator, I'd understand, but this seems to be doing work for nothing.

ViktorHofer replied:
Something to think about: What if CompareInfo in netfx suddenly uses the field for something else and (obviously) expects it to contain the correct value?

stephentoub replied:
> What if CompareInfo in netfx suddenly uses the field for something else and (obviously) expects it to contain the correct value?

Then that would be introducing a behavioral difference, which would a) break compatibility with existing versions, and b) be something we'd need to do in corefx as well.

ViktorHofer replied:
I was thinking more about a Windows-specific feature addition (b) like a new public method which a) wouldn't break compatibility with previous netfx versions (a). Because of (b) it wouldn't be ported to Core.

stephentoub replied:
> it wouldn't be ported to Core.

It absolutely would be.  We shouldn't be adding new netfx-only additions to existing types.

ViktorHofer commented:
> Can you explain this? I would expect not updating the blobs in CI to be a good thing, in that code changes still have to work with those known good blobs?

@danmosemsft To correct myself, we shouldn't update the blobs in CI but compare the canned base64 blob with ones that we dynamically create.

Example given:
We are running the tests in netcoreapp config. Then we should take the netcoreapp blob from the data and compare it with the during runetime serialized and base64 encoded blob of the respective object. Then we can guarantee that the blobs don't change by some changes in coreclr or corefx. I will add a test for that.

The current state of the PR with green CI is the perfect example of what I ment. We are still waiting for the CompareInfo change from coreclr to come in but as I already updated the blobs from my local build it's already green as it doesn't compare the real current state (besides the equality checks). So we could either refine all equality checks to really compare all possible fields (but what happens if another gets added?) or just compare the base64 blobs of the canned one with a just generated one.

krwq commented on src/System.Runtime.Serialization.Formatters/tests/BinaryFormatterHelper.cs:
```diff
@@ -184,22 +186,41 @@ public static void UpdateCoreTypeBlobs(string testDataFilePath, string[] blobs)
             Assert.Equal(numberOfBlobs, blobs.Length);
             File.WriteAllLines(testDataFilePath, updatedTestDataLines);
+
+            return (numberOfBlobs, numberOfFoundBlobs, numberOfUpdatedBlobs);
```
this additional checks are meaningless as long as this line is in place `if (!testDataLine.Trim().StartsWith("yield") || numberOfBlobs >= blobs.Length)`

krwq replied:
https://github.com/dotnet/corefx/pull/21223/commits/b32d9000a94c247f07436c17330634a96b1d9960#diff-ca8418a369f036135b158bf46069926eR183

ViktorHofer replied:
which checks do you mean? I haven't introduced any new checks, only new fields for returning more infos to the caller.

ViktorHofer replied:
> numberOfBlobs >= blobs.Length will skip blobs from the end

This is by design and favored by us. Right?

krwq replied:
yes, but because of that the numbers will pretty much always add up to what you expect them to be

ViktorHofer replied:
numberOfUpdatedBlobs is definitely interesting for the caller. numberofFoundBlobs is just a check if the regex pattern is working correctly and numberOfBlobs tells us how many active test cases we currently have.

krwq replied:
numberofFoundBlobs is capped by that checked I pointed you to

krwq replied:
what I mean is that it should count independently of that check and replacement logic - anyway do as you think is correct

danmosemsft commented:
New CLR
@dotnet-bot test this please

danmosemsft commented:
@ViktorHofer lots of "The stored blob for type System.Runtime.Serialization.Formatters.Tests.StructContainingOtherStructs is outdated and needs to be updated."

danmosemsft commented:
@dotnet/dnceng OSX has `Error during cleanup: java.io.IOException: remote file operation failed: /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest at hudson.remoting.Channel@9c15914:JNLP4-connect connection from 131.107.19.216/131.107.19.216:51720: hudson.remoting.ChannelClosedException: channel is already closed
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
Setting status of 70896de2c165037e9bdd813bbaa36f1bca20954d to FAILURE with url https://ci3.dot.net/job/dotnet_corefx/job/master/job/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/169/ and message: 'Build finished. '
Using context: OSX x64 Debug Build
java.nio.channels.ClosedChannelException
	at org.jenkinsci.remoting.protocol.impl.ChannelApplicationLayer.onReadClosed(ChannelApplicationLayer.java:208)
	at org.jenkinsci.remoting.protocol.ApplicationLayer.onRecvClosed(ApplicationLayer.java:222)
	at org.jenkinsci.remoting.protocol.ProtocolStack$Ptr.onRecvClosed(ProtocolStack.java:832)
	at org.jenkinsci.remoting.protocol.FilterLayer.onRecvClosed(FilterLayer.java:287)`

danmosemsft commented:
@joperezr I am still seeing Nano fail with `2017-06-21 17:21:14,918: INFO: proc(54): run_and_log_output: Output: This version of C:\dotnetbuild\work\365b7637-dddf-4b3e-8980-41f402301770\Payload\dotnet.exe is not compatible with the version of Windows you're running. Check your computer's system information and then contact the software publisher.` ...

ViktorHofer commented:
@dotnet-bot test this please

_Merged another clr maestro PR. This is not failing for me locally_

ViktorHofer commented:
Found the reason of the errors. In CI the assembly version of System.Runtime.Serialization.Formatters.Tests.dll where a few test types live is higher than locally (4.0.2.0 vs 4.0.3.0). I removed these static infos now for comparison and also found out that CookieCollection m_version field needs to be properly initialized to be present in the serialization payload. Also fixed now.

danmosemsft commented:
@ViktorHofer more failures

ViktorHofer commented:
Checking...

ViktorHofer commented:
What is currently happening here doesn't make any sense... The blobs are different in CI. E.g. for CookieCollection the stored blob contains the recently introduced field m_versions, in CI the runtime generated blob doesn't contain the field (verify-able with base64 decode). 

I see that different versions of assemblies are used locally vs CI. Is that expected, is CI really picking up the correct commits and compiling it to the correct assemblies as locally? ...

danmosemsft commented:
Try pushing the same changes up to ci in another branch to get confirmation

danmosemsft commented:
Btw might be helpful to change assert equals on the blobs to be on the decoded blobs to get nicer messages?

danmosemsft commented:
@dotnet/dnceng ci seems to be using old sources

danmosemsft commented:
@viktorhofer I meant, creating a fresh new pr in parallel

ViktorHofer commented:
The thing is that the decoded values could be the same which must not mean that the encoded ones are also the same. To understand errors you need the encoded ones. 

I know it's a pain to then copy it and paste it into a base64decode tool...

PS: The errors might be related to not being based on the latest changes in master. I did a rebase and push forced to verify that.

mmitche commented:
@danmosemsft What do you mean?

ViktorHofer commented:
@mmitche

> I see that different versions of assemblies are used locally vs CI. Is that expected, is CI really picking up the correct commits and compiling it to the correct assemblies as locally? ...

Maybe I'm doing something wrong. Not sure. Waiting for new CI tests

mmitche commented:
@ViktorHofer @danmosemsft Can you compare the checked out hashes (seen towards the top of the log in the git-output).

mmitche commented:
@ViktorHofer Which leg specifically?  It's picking up the "merged" hash, which is the commit which would be created if you merged this PR with master at this moment.  You can see the exact checked out hashes at the top.

mmitche commented:
Can someone on @dotnet/dnceng help out @ViktorHofer understand what the failures are?

I'm assuming Windows x64 Debug/Release is the leg you're referring to?

markwilkie commented:
@crummel can help take a look.

crummel commented:
@mmitche Yes, I'm looking at this now.

@ViktorHofer So you're thinking that the tests are picking up different test data files in CI vs. on your local machine?  I agree that Jenkins appears to be picking up the right files, I'm trying this locally now.

ViktorHofer commented:
I'm not sure, it might be the case as I see different assembly versions used in the base64 blobs which we are generated the System.Runtime.Serialization.Formatters tests. If you build it pocally, pls try also to run the mentioned tests. If you are also getting the same failures then it might be my machine which is acting weird.

ViktorHofer commented:
Tests are also failing in Linux and OSX legs, not just Windows. It might not be connected to CI, it was just a guess...

MattGal commented:
The PR touched System.Runtime.Serialization.Formatters and System.Runtime.Serialization.Formatters.Tests.BinaryFormatterTests are failing all over.   What about this looks like an infra issue?

danmosemsft commented:
@dotnet/dnceng hold off for now -- I've thrown up the same code in another PR and we'll see the results there. Thanks for looking.

crummel commented:
@ViktorHofer I got a similar test failure - I didn't check the actual base64 output but I put it at https://gist.github.com/crummel/47aec139efe0bdc7def94ba6d92bed43 if it's helpful for you.

danmosemsft commented:
@krwq I got similar results in my dupe run. It's not CI's fault :)  Can you please make this your first priority today as it's 2.0 work.

krwq commented:
I've updated the blobs - I was seeing some local failures about blobs being out of date. As a sanity check I changed culture inside the tests and rerun - all pass, also changed assembly version of a test project (properly we would need to do that globally) and also saw all tests passing with new blobs.

ViktorHofer commented:
Great... That means my local corefx build is useless and was the culprit -,- Sry for that!

ViktorHofer commented:
@krwq We also need the updated blobs in the release branch. Can you update that one too? I don't trust my local build anymore...

krwq commented:
@ViktorHofer no worries, those are fragile :-) Hopefully this will be consistent with CI. It would be nice to understand why it is not working on your box as the same thing could be hit by customers.

Could you try clean corefx build? (`git clean -fdx`)

krwq commented:
seems like we do have some differences with CI 😉

krwq commented:
@ViktorHofer I checked one random failure and here are my results.
The deserialized data is identical.
I cracked opened two blobs and compared them and see following differences:
- serialized data has an XML in it which has different line endings `\r\n` vs `\n` (majority of the diff)
- just before the xml I can see something which looks like is labelled as `System.Version`
`02 00 00 00 09 03 00 00 00 06 04 00 00 00 A6 09` (A6 = 166; A609=42505)
vs
`02 00 00 00 09 03 00 00 00 06 04 00 00 00 8E 09` (8E = 142; 8E09=36361)
currently not sure what that is but considering this has 16 bytes and System.Version has 4 fields it could be a revision field
```
        private readonly int _Major; // Do not rename (binary serialization)
        private readonly int _Minor; // Do not rename (binary serialization)
        private readonly int _Build = -1; // Do not rename (binary serialization)
        private readonly int _Revision = -1; // Do not rename (binary serialization)
```

I think we can safely ignore line endings since xml parsers are required to normalize them not sure about the version (also not sure what kind of version it is)

krwq commented:
@ViktorHofer, I talked with @danmosemsft offline and we believe the diff is ok. It would be nice to understand it more and fix but we do not believe this is blocking.
